src/coreComponents/physicsSolvers/multiphysics/poromechanicsKernels/ThermalMultiphasePoromechanics.hpp

```diff
@@ -0,0 +1,350 @@
+   */
+  ThermalMultiphasePoromechanics( NodeManager const & nodeManager,
+                                  EdgeManager const & edgeManager,
+                                  FaceManager const & faceManager,
+                                  localIndex const targetRegionIndex,
+                                  SUBREGION_TYPE const & elementSubRegion,
+                                  FE_TYPE const & finiteElementSpace,
+                                  CONSTITUTIVE_TYPE & inputConstitutiveType,
+                                  arrayView1d< globalIndex const > const inputDispDofNumber,
+                                  globalIndex const rankOffset,
+                                  CRSMatrixView< real64, globalIndex const > const inputMatrix,
+                                  arrayView1d< real64 > const inputRhs,
+                                  real64 const (&gravityVector)[3],
+                                  string const inputFlowDofKey,
+                                  localIndex const numComponents,
+                                  localIndex const numPhases,
+                                  string const fluidModelKey );
+
+  /**
+   * @class StackVariables
+   * @copydoc geosx::finiteElement::ImplicitKernelBase::StackVariables
+   *
+   * Adds a stack array for the displacement, incremental displacement, the
+   * constitutive stiffness, etc.
+   */
+  struct StackVariables : public Base::StackVariables
+  {
+public:
+
+    /// Constructor.
+    GEOSX_HOST_DEVICE
+    StackVariables():
+      Base::StackVariables()
+    {}
+
+    /// Derivative of mass increment wrt temperature
+    real64 dCompMassIncrement_dTemperature[maxNumComponents]{};
+    /// Derivative of pore volume constraint wrt temperature
+    real64 dPoreVolConstraint_dTemperature;
+    /// Derivative of body force wrt temperature
+    real64 dBodyForce_dTemperature[3]{};
+
+    /// Energy increment
+    real64 energyIncrement{};
+    /// Derivative of energy increment wrt volumetric strain increment
+    real64 dEnergyIncrement_dVolStrainIncrement{};
+    /// Derivative of energy increment wrt pressure
+    real64 dEnergyIncrement_dPressure{};
+    /// Derivative of energy increment wrt temperature
+    real64 dEnergyIncrement_dTemperature{};
+    /// Derivative of energy increment wrt global comp fraction
+    real64 dEnergyIncrement_dComponents[maxNumComponents]{};
+
+    // Storage for mass/energy residual and degrees of freedom
+
+    /// Derivative of linear momentum balance residual wrt temperature
+    real64 dLocalResidualMomentum_dTemperature[Base::StackVariables::numDispDofPerElem][1]{};
+    /// Derivative of mass balance residual wrt pressure
+    real64 dLocalResidualMass_dTemperature[maxNumComponents][1]{};
+    /// Derivative of pore volume constraint residual wrt pressure
+    real64 dLocalResidualPoreVolConstraint_dTemperature[1][1]{};
+
+    /// Energy balance residual
+    real64 localResidualEnergy[1]{};
```

Comment:
Why do you have those `[1]` here (and before and after)?

Reply:
Here, this is needed to call the `BilinearFormUtilities`. I know this is not great, but changing that would involve a lot of changes in the finite element kernels, which is beyond the scope of the PR. But yes, I agree with you that we should improve that at some point.

Reply:
In general we always treat local residuals as vectors and jacobians as matrices, even if they happen to be of size 1 (1x1) for particular physics.